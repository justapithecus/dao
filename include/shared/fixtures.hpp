#define APPROVALS_UT
#include "ApprovalTests.hpp"

using namespace boost::ut;
using namespace ApprovalTests;
