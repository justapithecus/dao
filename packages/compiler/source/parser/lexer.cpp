#include "../utils.hpp"

#include "lexer.hpp"
#include "state_machine.h"

namespace dao {

  auto lex(std::string_view fpath) -> std::vector<dao::token> {
    std::vector<dao::token> tokens{};
    dao::lexeme             lexeme{};
    auto                    source{dao::fread(fpath.data())};
    auto                    src_ptr{source.data()};
    unsigned char           state{lexical_state_next_char};

    do {
      unsigned char ch;
      auto          eqc = equivalence_class[(ch = *src_ptr++)];

      state = transition[state + eqc];
      src_ptr += ch_rewind[state];
      // lexeme.col_num += ch_rewind[state];
      lexeme.len += inside[state];

      switch (state) {
      case lexical_state_next_char: {
        ++lexeme.col_num;
        break;
      }
      case lexical_state_identifier: {
        lexeme.token_kind = token_kind_identifier;
        break;
      }
      case lexical_state_numeral: {
        lexeme.token_kind = token_kind_numeral;
        break;
      }
      case lexical_state_operator:
        lexeme.token_kind = token_kind_operator;
        [[fallthrough]];
      case lexical_state_identifier_end:
        [[fallthrough]];
      case lexical_state_numeral_end: {
        lexeme.update_repr(src_ptr, 0);
        lexeme.reset();
        tokens.emplace_back(lexeme.as_token());
        break;
      }
      default:
        break;
      }

    } while (state > lexical_state_final);

    return tokens;
  }
} // namespace dao
