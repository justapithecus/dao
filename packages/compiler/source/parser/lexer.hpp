#pragma once

#include <iostream>
#include <string_view>
#include <vector>

#include "token.hpp"

namespace dao {

  struct lexeme {
    std::string  repr;
    std::uint8_t len;
    std::uint8_t line_num;
    std::uint8_t col_num;

    explicit lexeme()
      : len{0}
      , line_num{1}
      , col_num{1} {
    }

    auto update_repr(char const *src_ptr, int end_offset) {
      auto first{src_ptr - len};
      auto last{first + (len + end_offset)};
      repr = std::string{first, last};
    }

    // reset trackers for next lexeme
    auto reset(int offset = 0) {
      col_num += (len + offset);
      len = 0;
    }

    auto as_token(dao::token_kind kind) {
      reset();
      return dao::token{repr, kind};
    }
  };

  static_assert(sizeof(dao::lexeme) <= 32);

  auto lex(std::string_view fpath) -> std::vector<token>;
} // namespace dao
