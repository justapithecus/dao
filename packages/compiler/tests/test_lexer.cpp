#include "fixtures.hpp"
#include "parser/lexer.hpp"

auto directory = Approvals::useApprovalsSubdirectory("golden-files");

auto main() -> int {
  "fibonacci"_test = []() {
    auto lexemes = dao::lex("examples/fibonacci.dao");
    Approvals::verifyAll("fibonacci.dao", lexemes);
  };
}
