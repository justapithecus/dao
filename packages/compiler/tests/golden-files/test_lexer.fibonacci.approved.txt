fibonacci.dao


[0] =      Keyword | function |
[1] =   Identifier | fibonacci |
[2] =    Separator | ( |
[3] =   Identifier | x |
[4] =    Separator | ) |
[5] =     New Line | \n |
[6] =      Keyword | if |
[7] =   Identifier | x |
[8] =     Operator | < |
[9] =      Numeral | 3 |
[10] =      Keyword | then |
[11] =     New Line | \n |
[12] =      Numeral | 1 |
[13] =     New Line | \n |
[14] =      Keyword | else |
[15] =     New Line | \n |
[16] =   Identifier | fibonacci |
[17] =    Separator | ( |
[18] =   Identifier | x |
[19] =     Operator | - |
[20] =      Numeral | 1 |
[21] =    Separator | ) |
[22] =     Operator | + |
[23] =   Identifier | fibonacci |
[24] =    Separator | ( |
[25] =   Identifier | x |
[26] =     Operator | - |
[27] =      Numeral | 2 |
[28] =    Separator | ) |
[29] =     New Line | \n |
[30] =     New Line | \n |
[31] =   Identifier | fibonacci |
[32] =    Separator | ( |
[33] =      Numeral | 40 |
[34] =    Separator | ) |
[35] =     New Line | \n |

